lib/handlers/common/GitHubCredentialsResolver.ts

```diff
@@ -33,58 +32,104 @@ import {
     ScmProviderByType,
 } from "../../typings/types";
 
+/**
+ * Type to implement different strategies to obtain a GitHub token
+ */
+type ObtainToken = (context: HandlerContext, id?: RemoteRepoRef) => Promise<string | undefined>;
+
 @Parameters()
 export class GitHubCredentialsResolver implements CredentialsResolver {
 
     @Secret(Secrets.OrgToken)
-    private readonly orgToken: string;
-
-    @Value({ path: "token", required: false, type: "string" })
-    private readonly clientToken: string;
+    public readonly orgToken: string;
 
     public async eventHandlerCredentials(context: HandlerContext,
                                          id?: RemoteRepoRef): Promise<ProjectOperationCredentials> {
-        return this.credentials(context, id);
+        return this.credentials(
+            [
+                obtainTokenFromConfiguration(this),
+                ObtainTokenFromIncomingMessage,
+                ObtainTokenFromProvider,
+            ],
+            context,
+            id);
     }
 
     public async commandHandlerCredentials(context: HandlerContext,
                                            id?: RemoteRepoRef): Promise<ProjectOperationCredentials> {
-        return this.credentials(context, id);
+        return this.credentials(
+            [
+                ObtainTokenFromIncomingMessage,
+                obtainTokenFromConfiguration(this),
+                ObtainTokenFromProvider,
+            ],
+            context,
+            id);
     }
 
-    private async credentials(context: HandlerContext,
+    private async credentials(obtainTokens: ObtainToken[],
+                              context: HandlerContext,
                               id?: RemoteRepoRef): Promise<ProjectOperationCredentials> {
 
-        // First try to obtain the token from the incoming event or command request
-        const actx: AutomationContextAware = context as any;
-        if (actx.trigger && actx.trigger.secrets) {
-            const secret = actx.trigger.secrets.find(s => s.uri === Secrets.OrgToken);
-            if (secret && hasToken(secret.value)) {
-                return { token: secret.value };
+        for (const obtainToken of obtainTokens) {
+            const token = await obtainToken(context, id);
+            if (hasToken(token)) {
+                return { token };
             }
         }
 
-        if (hasToken(this.orgToken)) {
-            return { token: this.orgToken };
-        } else if (hasToken(this.clientToken)) {
-            return { token: this.clientToken };
-        } else if (hasToken(configurationValue("token", "null"))) {
-            return { token: configurationValue<string>("token")};
-        } else if (hasToken(configurationValue("sdm.github.token", "null"))) {
-            return { token: configurationValue<string>("sdm.github.token")};
-        }
+        throw new Error("No GitHub token available! Please add a token to the SDM configuration at 'sdm.github.token' "
```

Comment:
Could this also be sent as a message using the HandlerContext?

Reply:
Not really. The clients deal with those errors in different ways. Eg there’s code that works by catching an error thrown by the resolver.